grafonnext/bin/mkschema.sh

```diff
@@ -0,0 +1,33 @@
+#!/usr/bin/env bash
+set -euo pipefail
+IFS=$'\n\t'
+
+if [ "$#" != 2 ]; then
+  echo "Usage: $(basename "$0") VERSION OUTPUTDIR"
+  exit 1
+fi
+
+VERSION="$1"
+OUTPUTDIR="$2"
+
+mkdir -p "$OUTPUTDIR/$VERSION"
+
+TEMPDIR=$(mktemp -d)
+
+trap "rm -rf $TEMPDIR" EXIT
+
+git clone --branch "$VERSION" --depth 1 https://github.com/grafana/grafana.git "$TEMPDIR"
+cd "$TEMPDIR"
+
+cd pkg/coremodel/dashboard
+thema lineage gen jsonschema > "$OUTPUTDIR/$VERSION/dashboard.json"
+cd -
+
+# Made an assumption here, generating jsonschema for these don't work.
+#for FILE in $(find public -name models.cue); do
+#    DIR=$(dirname "$FILE")
+#    NAME=$(basename "$DIR")
+#    cd "$DIR"
+#    thema lineage gen jsonschema > "$OUTPUTDIR/$VERSION/$NAME.json"
```

Comment:
there's gonna have to be more structure than just version + name - all the composable kinds declared in plugins also have a slot associated with them. Panel plugins have the `Panel` slot, and datasource plugins the `DSOptions` and `Query` slots.

Given that, how would you like the dir structure to look? i've handwaved over a Go import structure that looks like this (slightly updated from original doc for kind system):

```
  // Core native types importable from standard, stable path
  "github.com/grafana/grok/core/dashboard/v0"
 
  // Composable kinds from core plugins also importable from standard, stable paths
  "github.com/grafana/grok/composable/timeseries/panel/v0"
  // Generated types are separated into subdirs by composition slot
  "github.com/grafana/grok/composable/prometheus/query/v0"
  "github.com/grafana/grok/composable/prometheus/dsconfig/v0"
```

you needn't use the same kind-category words (core, composable), but flattening is probably unwise if only because these things don't logically exist in separate namespaces upstream, so they could clobber. Also, i suspect it'd be tricky to tell how to re-present each of them if you've lost context about what variety of thing you're looking at?

i'll make the go code just use the above convention, you can modify as desired

Reply:
From the Grafonnet perspective, I'd like to have 1-on-1 mapping between the generated JSON schemas and a Grafana release. If there are multiple schemas, then they would all be grouped under `schemas/<grafana version>/<unique name>.json`, for example:

- `schemas/v.9.2.3/panel.timeseries.json`
- `schemas/v.9.2.3/datasource.prometheus.query.json` (I assume this to fill the `target` attribute in a panel?)
- `schemas/v.9.2.3/datasource.prometheus.dsconfig.json`

Dependencies with JSON schema can be resolved with `$id` and `$ref`, not sure how much of that you want to support with Thema.

Reply:
The directory structure from the go script can be used, we can figure out which library pattern works as we work through this, perhaps exposing users to the internal vocabulary might not be so bad after all.

Reply:
> perhaps exposing users to the internal vocabulary might not be so bad after all.

I'm really glad you say that :) 

While there is a lot of internal detail with the schema pipelines that users should be shielded from, the names that i exposed in the paths for the go implementation here are ones that are supposed to be user-facing. The [words we've chosen](https://github.com/grafana/grafana/issues/58711) may not be the best, but there are precise logical boundaries they correspond to. Those logical boundaries are learnable (though we don't yet have the docs to make it so), and learning them should make any as-code user much more confident about how all the pieces fit together.